File: backend/app/core/security.py
import base64
import hashlib
import hmac
import json
import time
from secrets import compare_digest
from fastapi import HTTPException, Request, status
from .config import settings


def issue_admin_token() -> str:
    """Create a short-lived, signed session token. The master key is never reused as a token."""
    payload = json.dumps({"scope": "admin", "exp": int(time.time()) + settings.admin_session_ttl_seconds}, separators=(",", ":")).encode()
    encoded = base64.urlsafe_b64encode(payload).rstrip(b"=")
    signature = hmac.new(settings.admin_api_key.encode(), encoded, hashlib.sha256).digest()
    return f"{encoded.decode()}.{base64.urlsafe_b64encode(signature).rstrip(b'=').decode()}"


def require_admin(request: Request):
    authorization = request.headers.get("Authorization", "")
    scheme, _, token = authorization.partition(" ")
    if scheme.lower() != "bearer" or not token:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Admin authentication is required")
    try:
        encoded, provided_signature = token.split(".", 1)
        expected_signature = base64.urlsafe_b64encode(
            hmac.new(settings.admin_api_key.encode(), encoded.encode(), hashlib.sha256).digest()
        ).rstrip(b"=").decode()
        payload = json.loads(base64.urlsafe_b64decode(encoded + "=" * (-len(encoded) % 4)))
    except (ValueError, UnicodeDecodeError, json.JSONDecodeError):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid admin session")
    if not compare_digest(provided_signature, expected_signature) or payload.get("scope") != "admin" or not isinstance(payload.get("exp"), int) or payload["exp"] < time.time():
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired admin session")

File: backend/src/services/security.js
const crypto = require("crypto");
const { UnauthorizedError } = require("../errors");

function base64UrlEncode(input) {
  return Buffer.from(input).toString("base64url");
}

function safeCompare(left, right) {
  const a = Buffer.from(String(left));
  const b = Buffer.from(String(right));
  if (a.length !== b.length) {
    return false;
  }
  return crypto.timingSafeEqual(a, b);
}

function issueScopedToken({ scope, expiresInSeconds, secret, claims = {} }) {
  const payload = {
    scope,
    exp: Math.floor(Date.now() / 1000) + expiresInSeconds,
    ...claims,
  };
  const encoded = base64UrlEncode(JSON.stringify(payload));
  const signature = crypto
    .createHmac("sha256", secret)
    .update(encoded)
    .digest("base64url");
  return `${encoded}.${signature}`;
}

function verifyScopedToken(token, { expectedScope, secret, errorMessage }) {
  const [encoded, providedSignature] = String(token || "").split(".");
  if (!encoded || !providedSignature) {
    throw new UnauthorizedError(errorMessage);
  }

  const expectedSignature = crypto
    .createHmac("sha256", secret)
    .update(encoded)
    .digest("base64url");

  if (!safeCompare(providedSignature, expectedSignature)) {
    throw new UnauthorizedError(errorMessage);
  }

  let payload;
  try {
    payload = JSON.parse(Buffer.from(encoded, "base64url").toString("utf8"));
  } catch {
    throw new UnauthorizedError(errorMessage);
  }

  if (
    payload.scope !== expectedScope ||
    typeof payload.exp !== "number" ||
    payload.exp < Math.floor(Date.now() / 1000)
  ) {
    throw new UnauthorizedError(errorMessage);
  }

  return payload;
}

function issueAdminToken(config) {
  return issueScopedToken({
    scope: "admin",
    expiresInSeconds: config.adminSessionTtlSeconds,
    secret: config.adminApiKey,
    claims: { subject: "admin" },
  });
}

function verifyAdminToken(token, config) {
  return verifyScopedToken(token, {
    expectedScope: "admin",
    secret: config.adminApiKey,
    errorMessage: "Invalid or expired admin session",
  });
}

function issueVolunteerToken(config, volunteerName) {
  const resolvedName = String(volunteerName || "Volunteer").trim() || "Volunteer";
  return issueScopedToken({
    scope: "volunteer_checkin",
    expiresInSeconds: config.volunteerSessionTtlSeconds,
    secret: config.volunteerTokenSecret,
    claims: {
      volunteer_name: resolvedName,
    },
  });
}

function verifyVolunteerToken(token, config) {
  return verifyScopedToken(token, {
    expectedScope: "volunteer_checkin",
    secret: config.volunteerTokenSecret,
    errorMessage: "Invalid or expired volunteer session",
  });
}

function requireAdmin(config) {
  return (req, _res, next) => {
    try {
      const authorization = req.header("authorization") || "";
      const [scheme, token] = authorization.split(" ");
      if ((scheme || "").toLowerCase() !== "bearer" || !token) {
        throw new UnauthorizedError("Admin authentication is required");
      }
      req.adminSession = verifyAdminToken(token, config);
      next();
    } catch (error) {
      next(error);
    }
  };
}

function requireVolunteer(config) {
  return (req, _res, next) => {
    try {
      const authorization = req.header("authorization") || "";
      const [scheme, token] = authorization.split(" ");
      if ((scheme || "").toLowerCase() !== "bearer" || !token) {
        throw new UnauthorizedError("Volunteer authentication is required");
      }
      const payload = verifyVolunteerToken(token, config);
      req.volunteerSession = payload;
      next();
    } catch (error) {
      next(error);
    }
  };
}

module.exports = {
  safeCompare,
  issueAdminToken,
  verifyAdminToken,
  issueVolunteerToken,
  verifyVolunteerToken,
  requireAdmin,
  requireVolunteer,
};
